File: 0001-0500/0290.Word Pattern.rb
# @param {String} pattern
# @param {String} str
# @return {Boolean}
def word_pattern(pattern, str)
    str = str.split(" ")
    if str.size == pattern.size
        p2s = Hash.new
        s2p = Hash.new
        for i in (0...pattern.size) do
            return false if !checkOrSet(pattern[i], str[i], p2s) or !checkOrSet(str[i], pattern[i], s2p)
        end
        true
    else
        false
    end
end
    
def checkOrSet(from, to, hash)
    if hash.include? from
        return false if hash[from] != to
    else
        hash[from] = to
    end
    return true
end

File: 0001-0500/0290.Word Pattern.py
"""
Result:
  29 / 29 test cases passed.
  Status: Accepted
  Runtime: 52 ms
  Your runtime beats 7.21% of pythonsubmissions.
Date:
  5/25/2016
"""
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        p2s = dict()
        s2p = dict()
        l = str.split()
        if len(pattern)!=len(l):
            return False
        for p,s in zip(pattern,l):
            b1 = p in p2s
            b2 = s in s2p
            if b1 and b2:
                if p2s[p]==s and s2p[s]==p:
                    continue
                else:
                    return False
            elif not b1 and not b2:
                p2s[p]=s
                s2p[s]=p
            else:
                return False
        return True
